caffe2/core/blob_serialization.cc

```diff
@@ -520,7 +526,14 @@ void TensorDeserializer::Deserialize(const TensorProto& proto, Tensor* tensor) {
                 (i + chunkBegin) * temp_blob.meta().itemsize(),
             1);
       }
-    }
+    } break;
+    case TensorProto_DataType_SPECIAL: {
+      CAFFE_THROW("SPECIAL Tensor is not handled yet.");
+    } break;
+    case TensorProto_DataType_NO_CONTENT: {
+      CAFFE_THROW("NO_CONTENT Tensor should not be deserialized.");
+    } break;
+      // Note: we intentially do not provide "default:" so if any new data types
```

Comment:
I might be wrong, but this is pretty questionable from the forward compatibility perspective unless you handle it in some other place. In the current version you'll be able silently load the model even if it's not possible to use it in current binary.

This is probably more like a question to @Yangqing around the original code in the first place

Reply:
This actually uses a compiler feature "-Wswitch", which I quote here:

-Wswitch
Warn whenever a switch statement has an index of enumerated type and lacks a case for one or more of the named codes of that enumeration. (The presence of a default label prevents this warning.) case labels outside the enumeration range also provoke warnings when this option is used (even if there is a default label). This warning is enabled by -Wall.

Alternative approach is to put a default and always fail explicitly - that might be safer indeed.

Reply:
Just in case my comment is primarily is about loading models produced by new version of the framework in the old binary.

-Wswitch will capture compilation errors when new type is added and not covered by switch, but even in this case models that have some new type like TensorProto_DataType_NO_CONTENT will be loaded by existing binaries with Caffe2 that have no idea how to handle this option and will just completely ignore it in switch, which will result in some unexpected behavior when you try to evaluate this model after loading.

Reply:
To keep the forward compatibility, we should add proto/IR version. Each time we update the proto, we bump up the version. And the runtime rejects the model, if the IR version is too high. Also for the NO_CONTENT option, we think I should move it to STORAGE_TYPE, because it's safer and looks more natural.

Reply:
So I will just leave this logic as it is now, and we can discuss further